include/algorithms/synthesis/dd_synthesis.hpp

```diff
@@ -1,9 +1,56 @@
 #pragma once
 
+#include "QuantumComputation.hpp"
 #include "core/truthTable/truth_table.hpp"
+#include "dd/FunctionalityConstruction.hpp"
+
+#include <chrono>
 
 namespace syrec {
 
     auto buildDD(const TruthTable& tt, std::unique_ptr<dd::Package<>>& dd) -> dd::mEdge;
 
+    class DDSynthesizer {
+    public:
+        DDSynthesizer() = default;
+
+        explicit DDSynthesizer(std::size_t nqubits):
+            qc(nqubits) {}
+
+        auto pathFromSrcDst(dd::mNode* src, dd::mNode const* dst, TruthTable::Cube::Vector& sigVec, TruthTable::Cube& tempVec) const -> void;
+
+        auto pathSignature(dd::mNode* src, TruthTable::Cube::Vector& sigVec, TruthTable::Cube& tempVec) -> void;
+
+        static auto terminate(dd::mEdge const& current) -> bool;
+
+        auto unifyPath(dd::mEdge& src, dd::mEdge const& current, TruthTable::Cube::Vector const& p1SigVec, TruthTable::Cube::Vector const& p2SigVec, const std::vector<std::size_t>& indices, std::unique_ptr<dd::Package<>>& dd) -> bool;
+
+        auto swapPaths(dd::mEdge& src, dd::mEdge const& current, std::unique_ptr<dd::Package<>>& dd) -> bool;
+
+        auto shiftUniquePaths(dd::mEdge& src, dd::mEdge const& current, std::unique_ptr<dd::Package<>>& dd) -> bool;
+
+        auto shiftingPaths(dd::mEdge& src, dd::mEdge const& current, std::unique_ptr<dd::Package<>>& dd) -> bool;
+
+        auto synthesize(dd::mEdge& src, std::unique_ptr<dd::Package<>>& dd) -> void;
+
+        auto               begin() noexcept { return qc.begin(); }
+        [[nodiscard]] auto begin() const noexcept { return qc.begin(); }
+        auto               end() noexcept { return qc.end(); }
+        [[nodiscard]] auto end() const noexcept { return qc.end(); }
```

Comment:
To me, it does not make much sense to have these functions here. The goal of the synthesizer is to produce a circuit from a DD. Iteration methods on the resulting circuit are provided by the QuantumComputation class itself.

Reply:
done 👍🏻